File: merge-intervals/insert-interval.py
from interval import *


def insert_interval(existing_intervals, new_interval):
    output = []
    for i, curr in enumerate(existing_intervals):
        if new_interval.end < curr.start:
            output.append(new_interval)
            return output + existing_intervals[i:]
        elif new_interval.start > curr.end:
            output.append(curr)
        else:
            new_interval = Interval(
                min(new_interval.start, curr.start), max(new_interval.end, curr.end)
            )

    output.append(new_interval)
    return output

File: merge-intervals/insert-interval.js
import { Interval } from "./interval.js";

export function insertInterval(existingIntervals, newInterval) {
	let result = [];

	for (let i = 0; i < existingIntervals.length; i++) {
		const currInterval = existingIntervals[i];

		if (newInterval.end < currInterval.start) {
			result.push(newInterval);
			return result.concat(existingIntervals.slice(i));
		} else if (newInterval.start > currInterval.end) {
			result.push(currInterval);
		} else {
			newInterval = new Interval(
				Math.min(newInterval.start, currInterval.start),
				Math.max(newInterval.end, currInterval.end)
			);
		}
	}

	result.push(newInterval);
	return result;
}
